### ISSUE TITLE
error when templating key-of from another templated param, InvalidDocblock affecting expected templated param count

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#1698]
# issue
* psalm seems to have less of an issue when defining DATA as a simple `array`, but this leads `key-of` to resolve to `array-key`, which then causes issues with `@return ID` on a method that returns `string`.
* `InvalidDocblock` seems to mess with the expected templated param count, I'm wondering if this would be mitigated by resolving "invalid" template to `as mixed` to suppress `TooManyTemplateParams` ?
* `MismatchingDocblockReturnType` is raised seemingly directly as a result of `key-of<DATA>` failing to parse.

as always, there's always the possibility that I'm grossly misusing psalm :P

# repro
* https://psalm.dev/r/0135f229d6
* using b19d256

```php
<?php
/**
* @template DATA as array<string, scalar|array|object|null>
* @template ID_VAL as array-key
*/
abstract class Foo {
  /**
  * @var DATA
  */
  protected $data;
  
  /**
  * @param DATA $data
  */
  public function __construct(array $data) {
    $this->data = $data;
  }
  
  /**
  * @template K as key-of<DATA>
  *
  * @param K $property
  *
  * @return DATA[K]
  */
  public function __get(string $property) {
    return $this->data[$property];
  }
  
  /**
  * @return ID_VAL
  */
  public function ObtainRepoId() {
    return $this->__get(static::IdProperty());
  }
  
  /**
  * @return key-of<DATA>
  */
  abstract public static function IdProperty() : string;
}

/**
* @template-extends Foo<array{id:int, name:string}, 'id', int>
*/
class Bar extends Foo {
  public static function IdProperty() : string {
    return 'id';
  }
}


/**
* @template-extends Foo<array{id:int, name:string}, 'id', int>
*/
class Baz extends Foo {
  public static function IdProperty() : string {
    return 'id';
  }
}
```
> Psalm output (using commit b19d256): 
> INFO: MixedArrayTypeCoercion - 27:12 - Coercion from array offset type 'K as key-of<DATA>' to the expected type 'string'
> ERROR: InvalidReturnStatement - 27:12 - The type 'array<array-key, mixed>|scalar|object|null' does not match the declared return type 'DATA[K]' for Foo::__get
> ERROR: InvalidReturnType - 24:13 - The declared return type 'DATA[K]' for Foo::__get is incorrect, got 'array<array-key, mixed>|scalar|object|null'
> INFO: MixedArgumentTypeCoercion - 34:25 - Argument 1 of Foo::__get expects string, parent type key-of<DATA> provided
> INFO: MixedReturnStatement - 34:12 - Could not infer a return type
> INFO: MixedInferredReturnType - 31:13 - Could not verify return type 'ID_VAL' for Foo::ObtainRepoId
> ERROR: MismatchingDocblockReturnType - 38:50 - Docblock has incorrect return type 'key-of<DATA>', should be 'string'
> ERROR: TooManyTemplateParams - 46:7 - Bar has too many template params, expecting 2
> ERROR: MismatchingDocblockReturnType - 47:41 - Docblock has incorrect return type 'key-of<DATA>', should be 'string'
> INFO: MixedInferredReturnType - 47:41 - Could not verify return type 'key-of<DATA>' for Bar::IdProperty
> ERROR: TooManyTemplateParams - 56:7 - Baz has too many template params, expecting 2
> ERROR: MismatchingDocblockReturnType - 57:41 - Docblock has incorrect return type 'key-of<DATA>', should be 'string'
> INFO: MixedInferredReturnType - 57:41 - Could not verify return type 'key-of<DATA>' for Baz::IdProperty

**Comment by muglug** [comment]
`@template K as key-of<T>` shouldn't be defined at the class level - I've updated the example above to reflect proper usage, but there are still some bugs evident.

**Comment by SignpostMarv** [comment]
*edited to update the error list & commit ref

**Comment by SignpostMarv** [comment]
> `@template K as key-of<T>` shouldn't be defined at the class level - I've updated the example above to reflect proper usage, but there are still some bugs evident.

the idea with this was so that psalm wouldn't have to make a round trip to verify that `@template ID as key-of<T>` was present on `@template T as array<string, scalar|array|object|null>`, with the implicit linkage between `@return ID` in `Foo::IdProperty()`, `@return ID_VAL` on  `Foo::ObtainRepoId()` with the method body passing `Foo::IdProperty()` to `Foo::__get()` implicitly hinting that `ID_VAL` is the specific `value-of<T>` corresponding to `ID as key-of<T>`. While the updated example is cleaner (i.e. DRY), it *feels* a little looser?

**Comment by muglug** [comment]
Cleaned-up example showing error as desired:

https://psalm.dev/r/4868c501f0
